src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/ReferenceAssemblies.cs

```diff
@@ -460,14 +460,51 @@ private async Task<ImmutableArray<MetadataReference>> ResolveCoreAsync(string la
                     }
                 }
 
-                // Prefer assemblies from the reference assembly package to ones otherwise provided
+                // Prefer newer assemblies when more than one have the same name
                 if (ReferenceAssemblyPackage is not null)
                 {
-                    var referenceAssemblies = new HashSet<string>(resolvedAssemblies.Where(resolved => resolved.StartsWith(referenceAssemblyInstalledPath!)));
+                    var comparer = new FrameworkPrecedenceSorter(DefaultFrameworkNameProvider.Instance, allEquivalent: false);
+                    var assembliesByName = resolvedAssemblies.GroupBy(Path.GetFileNameWithoutExtension, StringComparer.OrdinalIgnoreCase);
+                    var assembliesToRemove = new List<string>();
+                    foreach (var assemblyNameGroup in assembliesByName)
+                    {
+                        var assembliesByPrecedence = assemblyNameGroup.OrderBy(GetFrameworkNameFromPath, comparer).ThenByDescending(GetFrameworkNameFromPath, new NuGetFrameworkSorter()).ToArray();
+                        for (var i = 1; i < assembliesByPrecedence.Length; i++)
+                        {
+                            // We want to keep the last reference listed for the most recent supported target framework.
+                            // Typically, if more than one item has the most recent supported target framework, it will
+                            // be a case where the reference assembly package provides the assembly and a newer version
+                            // is provided explicitly. For example:
+                            //
+                            // Microsoft.NETCore.App.Ref 6.0.0 provides System.Collections.Immutable in the net6.0 folder
+                            // System.Collections.Immutable 8.0.0 provides System.Collections.Immutable in the net6.0 folder
+                            //
+                            // In this example, the Microsoft.NETCore.App.Ref package is resolved first, so by taking
+                            // the last net6.0 assembly, we ensure the assembly from System.Collections.Immutable 8.0.0
+                            // is resolved.
+                            if (Equals(GetFrameworkNameFromPath(assembliesByPrecedence[0]), GetFrameworkNameFromPath(assembliesByPrecedence[i])))
```

Comment:
>GetFrameworkNameFromPath(assembliesByPrecedence[0]) [](http://example.com/codeflow?start=39&length=51)

as this is invariant, consider moving it outside the loop

Reply:
➡️ Moved to single evaluation of `GetFrameworkNameFromPath`